torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -758,147 +773,25 @@ def sort_tokens(self, x, topk_ids, topk_weights):
-                    hidden_outputs,
-                    w_down.transpose(-2, -1),
-                    m_offsets,
-                    out_dtype=torch.bfloat16,
-                )
-
-            return hidden_outputs
-
-        elif self.group_mm == "ds":
-            # TODO - ds gg doesn't like the ending padding, so I currently lift only valid tokens out.
-            # This needs to be fixed as cpu sync and also breaks cuda graph by overwriting memory at the last update.
-            valid_tokens = contig_tokens[: m_offsets[-1]]
-            # this function creates the required m_indices for dsgemm while ensuring no cpu-gpu sync.
-            m_indices = dsgemm_utils.create_indices_from_offsets_nosync(m_offsets)
-
-            # Get expert weights for all projections
-            gate_proj_weight_fp8 = self.get_parameter("gate_proj_fp8")
-            gate_proj_scales = self.get_parameter("gate_proj_scales")
-            up_proj_weight_fp8 = self.get_parameter("up_proj_fp8")
-            up_proj_scales = self.get_parameter("up_proj_scales")
-            down_proj_weight_fp8 = self.get_parameter("down_proj_fp8")
-            down_proj_scales = self.get_parameter("down_proj_scales")
-
-            # Get dimensions
-            m_actual_tokens = valid_tokens.shape[0]
-            intermediate_size = self.get_parameter("gate_proj_weight").shape[1]
-            hidden_size = self.get_parameter("down_proj_weight").shape[1]
-
-            # Prepare input in FP8 format (shared by gate and up projections)
-            if self.use_triton_quant:
-                gate_up_input = dsgemm_kernels.groupwise_activation_quant(
-                    valid_tokens
-                )  # dsgemm_kernels.activation_quant_triton(valid_tokens)
-            else:
-                gate_up_input = dsgemm_utils.prepare_fp8_input(valid_tokens)
 
-            # Allocate output buffers
-            gate_proj_out = torch.empty(
-                (m_actual_tokens, intermediate_size),
-                device=contig_tokens.device,
-                dtype=contig_tokens.dtype,
-            )
-            up_proj_out = torch.empty_like(gate_proj_out)
-
-            # Run first GEMM (gate projection)
-            deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_contiguous(
-                gate_up_input,
-                (gate_proj_weight_fp8, gate_proj_scales),
-                gate_proj_out,
-                m_indices,
-            )
+    def _run_group_gemm(self, contig_tokens, m_sizes, m_offsets):
+        """Run the appropriate group GEMM implementation based on configuration"""
 
-            # Run second GEMM (up projection)
-            deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_contiguous(
-                gate_up_input,
-                (up_proj_weight_fp8, up_proj_scales),
-                up_proj_out,
-                m_indices,
+        try:
+            return self.group_gemm_strategies[self.group_mm].execute(
+                contig_tokens, m_sizes, m_offsets, self
             )
```

Comment:
What is the last `self` argument for?

Reply:
it's passing in the current module (i.e. MoE) so that the groupgemm strategy can have access to the expert weights via module.